elks/arch/i86/drivers/char/console-direct.c

```diff
@@ -183,54 +246,114 @@ struct tty_ops dircon_ops = {
+    for (i = 0; i < p->vseg_bytes; i += 2)
+        pokew(i, p->vseg_base, 0x5555);
+    for (i = 0; i < p->vseg_bytes; i += 2)
+        if (peekw(i, p->vseg_base) != 0x5555) return 1;
+    for (i = 0; i < p->vseg_bytes; i += 2)
+        pokew(i, p->vseg_base, 0x7 << 8 | ' ');
+
+    /* Enable video */
+    outb(0x09, p->crtc_base + CRTC_MODE);
+    return 0;
+}
 
-    for (i = 0; i < NumConsoles; i++) {
-        C->cx = C->cy = 0;
-        if (!i) {
-            C->cx = peekb(0x50, 0x40);
-            C->cy = peekb(0x51, 0x40);
-        }
-        C->fsm = std_char;
-        C->basepage = i * PageSizeW;
-        C->vseg = VideoSeg + (C->basepage >> 3);
-        C->attr = A_DEFAULT;
+static const char *type_string(enum OutputType t)
+{
+    switch (t) {
+    case OT_MDA: return "MDA";
+    case OT_CGA: return "CGA";
+    case OT_EGA: return "EGA";
+    case OT_VGA: return "VGA";
+    }
+}
 
+void INITPROC console_init(void)
+{
+    Console *C;
+    int i, j, dev;
+    unsigned short boot_crtc;
+    enum OutputType boot_type;
+    unsigned char screens = 0;
+    unsigned char cur_output = 0;
+    boot_crtc = peekw(0x63, 0x40);
+    for (i = 0; i < N_DEVICETYPES; ++i)
+        if (params[i].crtc_base == boot_crtc) boot_type = i;
+
+    C = &Con[0];
+    for (i = 0; i < N_DEVICETYPES; ++i) {
+        dev = (i + boot_type) % N_DEVICETYPES;
+        if (!probe_crtc(params[dev].crtc_base)) continue;
+        params[dev].is_present = 1;
+        screens++;
+        init_output(dev);
+        for (j = 0; j < params[dev].max_pages; ++j) {
+            C->cx = C->cy = 0;
+            C->op = cur_output;
+            if (!j) Visible[C->op] = C;
+            if (!j && !i) {
+                C->cx = peekb(0x50, 0x40);
+                C->cy = peekb(0x51, 0x40);
+            }
+            C->fsm = std_char;
+            C->vseg_offset = j * params[dev].page_words;
+            C->vseg = params[dev].vseg_base + (C->vseg_offset >> 3);
+            C->attr = A_DEFAULT;
+            C->type = dev;
+            C->w = params[dev].w;
+            C->max_col = C->w - 1;
+            C->h = params[dev].h;
+            C->max_row = C->h - 1;
+            C->crtc_base = params[dev].crtc_base;
 #ifdef CONFIG_EMUL_ANSI
-        C->savex = C->savey = 0;
+            C->savex = C->savey = 0;
 #endif
-
-        /* Do not erase early printk() */
-        /* ClearRange(C, 0, C->cy, MaxCol, MaxRow); */
-
-        C++;
+            NumConsoles++;
+            if (i) DisplayCursor(C, 0);
+            C++;
+        }
+        cur_output++;
     }
 
     kbd_init();
-
-    printk("Direct console, %s kbd %ux%u"TERM_TYPE"(%d virtual consoles)\n",
-           kbd_name, Width, Height, NumConsoles);
+    printk("boot_crtc: %x\n", Con[0].crtc_base);
+    printk("Direct console %s kbd"TERM_TYPE"(%d screens):\n", kbd_name, screens);
+    for (i = 0; i < NumConsoles; ++i) {
+        printk("/dev/tty%i, %s, %ux%u\n", i + 1, type_string(Con[i].type), Con[i].w, Con[i].h);
```

Comment:
We can't add more printk lines to the startup boot screen, as it then scrolls past and limits determining what has gone wrong. Suggest moving all extra info inside monitor-ibmpc.c, where if the option is enabled that is less of an issue.

In summary, I would say lets try to keep console-*.c, and especially console-direct.c as little modified as possible, with all the hardware related stuff moved to a new .c file. That has the benefit of leaving the system as close to as it is now when not configured, but allowing you to shoot high when running your configured code. This is because this is likely a very specialized configuration option, after all.

Reply:
That `boot_crtc: ` message was more for debugging, I'll remove that. Then the rest of that can be squeezed back down to that single line that was there before.

Reply:
> Then the rest of that can be squeezed back down to that single line that was there before.

Don't get me wrong @vkoskiv, I encourage you to write lots of fun code, have debug messages and try new things. We have a `debug()` macro system in debug.h that is pretty fancy for optionally leaving debug messages intact but not displayed until wanted

In general, if we can get this somewhat complicated PR separating most of your new hardware (and other init?) code into a separate file, then it's OK with me to use a multiple-PR approach to adding or perfecting features. That kind of lets you have your own file to play/experiment/enhance, while keeping the rest of the system stable and relatively unmodified. In that way, you can join the fun and write some interesting code while not having to worry about the state/status/direction of the entire codebase, especially in the beginning.

Just wanted to clarify that! :)